blas/impl/KokkosBlas1_axpby_unification_attempt_traits.hpp

```diff
@@ -0,0 +1,955 @@
+struct getLayoutFromView {
+  using type = void;
+};
+
+template <typename T>
+struct getLayoutFromView<T, true> {
+  using type = typename T::array_layout;
+};
+
+// --------------------------------
+
+template <class tExecSpace, class AV, class XMV, class BV, class YMV>
+struct AxpbyUnificationAttemptTraits {
+  // ********************************************************************
+  // Terminology:
+  // - variable names begin with lower case letters
+  // - type names begin with upper case letters
+  // ********************************************************************
+ private:
+  static constexpr bool onDevice =
+      KokkosKernels::Impl::kk_is_gpu_exec_space<tExecSpace>();
+  static constexpr bool onHost = !onDevice;
+
+  static constexpr bool a_is_scalar = !Kokkos::is_view_v<AV>;
+  static constexpr bool a_is_r0     = Tr0_val<AV>();
+  static constexpr bool a_is_r1s    = Tr1s_val<AV>();
+  static constexpr bool a_is_r1d    = Tr1d_val<AV>();
+
+  static constexpr bool x_is_r1 = Kokkos::is_view_v<XMV> && (XMV::rank == 1);
+  static constexpr bool x_is_r2 = Kokkos::is_view_v<XMV> && (XMV::rank == 2);
+
+  static constexpr bool b_is_scalar = !Kokkos::is_view_v<BV>;
+  static constexpr bool b_is_r0     = Tr0_val<BV>();
+  static constexpr bool b_is_r1s    = Tr1s_val<BV>();
+  static constexpr bool b_is_r1d    = Tr1d_val<BV>();
+
+  static constexpr bool y_is_r1 = Kokkos::is_view_v<YMV> && (YMV::rank == 1);
+  static constexpr bool y_is_r2 = Kokkos::is_view_v<YMV> && (YMV::rank == 2);
+
+  static constexpr bool xyRank1Case = x_is_r1 && y_is_r1;
+  static constexpr bool xyRank2Case = x_is_r2 && y_is_r2;
+
+  // ********************************************************************
+  // Declare 'AtInputScalarTypeA_nonConst'
+  // ********************************************************************
+  using ScalarTypeA2_onDevice =
+      typename getScalarTypeFromView<AV, (a_is_r0 || a_is_r1s || a_is_r1d) &&
+                                             onDevice>::type;
+  using ScalarTypeA1_onDevice =
+      std::conditional_t<a_is_scalar && onDevice, AV, ScalarTypeA2_onDevice>;
+
+  using ScalarTypeA2_onHost =
+      typename getScalarTypeFromView<AV, (a_is_r0 || a_is_r1s || a_is_r1d) &&
+                                             onHost>::type;
+  using ScalarTypeA1_onHost =
+      std::conditional_t<a_is_scalar && onHost, AV, ScalarTypeA2_onHost>;
+
+  using AtInputScalarTypeA =
+      std::conditional_t<onHost, ScalarTypeA1_onHost, ScalarTypeA1_onDevice>;
+
+  using AtInputScalarTypeA_nonConst =
+      typename std::remove_const<AtInputScalarTypeA>::type;
+
+  // ********************************************************************
+  // Declare 'AtInputScalarTypeX_nonConst'
+  // ********************************************************************
+  using AtInputScalarTypeX = typename XMV::value_type;
+
+  using AtInputScalarTypeX_nonConst =
+      typename std::remove_const<AtInputScalarTypeX>::type;
```

Comment:
Note: Kokkos::View defines the type `Kokkos::View::non_const_value_type` so no need to call `std::remove_const`

Reply:
Thanks for the hint. Done here and in another similar point.